components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/AbstractHTTPFunction.java

```diff
@@ -73,65 +89,298 @@ public AbstractHTTPFunction() {
         allowedDomains = ConfigProvider.getInstance().getAllowedDomainsForHttpFunctions();
     }
 
-    protected void executeHttpMethod(HttpUriRequest request, Map<String, Object> eventHandlers) {
+    protected void executeHttpMethod(HttpUriRequest clientRequest, Map<String, Object> eventHandlers,
+                                     AuthConfigModel authConfigModel) {
 
         AsyncProcess asyncProcess = new AsyncProcess((context, asyncReturn) -> {
-            JSONObject json = null;
-            int responseCode;
             String outcome;
             String endpointURL = null;
 
-            if (request.getURI() != null) {
-                endpointURL = request.getURI().toString();
+            HttpUriRequest request;
+            try {
+                if (authConfigModel != null) {
+                    AuthConfig authConfig = AuthConfigFactory.getAuthConfig(authConfigModel, context, asyncReturn);
+                    request = authConfig.applyAuth(clientRequest, authConfigModel);
+                } else {
+                    request = clientRequest;
+                }
+
+                if (request.getURI() != null) {
+                    endpointURL = request.getURI().toString();
+                }
+
+                if (!isValidRequestDomain(request.getURI())) {
+                    LOG.error("Provided Url does not contain a allowed domain. Invalid Url: " + endpointURL);
+                    asyncReturn.accept(context, Collections.emptyMap(), Constants.OUTCOME_FAIL);
+                } else {
+                    Pair<Integer, Pair<String, JSONObject>> result = executeRequest(request, endpointURL);
+                    if (result.getLeft() != HTTP_STATUS_TOO_MANY_REQUESTS &&
+                            result.getLeft() != HTTP_STATUS_UNAUTHORIZED &&
+                            !result.getRight().getLeft().equals(Constants.OUTCOME_SUCCESS)) {
+                        LOG.error("Error while calling endpoint. Url: " + endpointURL);
+                        result = executeRequestWithRetries(request, endpointURL, maxRequestAttemptsForAPIEndpointTimeout);
+                    }
+                    outcome = result.getRight().getLeft();
+                    JSONObject json = result.getRight().getRight();
+                    asyncReturn.accept(context, json != null ? json : Collections.emptyMap(), outcome);
```

Comment:
Rather than retrying if specific status codes are met I think we have to retry if it's not in the range of 200 to 300 -> success cases, 300 to 400 -> redirects, 400 to 500 -> client errors

Reply:
Resolved with https://github.com/wso2-extensions/identity-conditional-auth-functions/pull/157/commits/778ee19b0b9677f6d2cf96e829414272e804de01